File: app-macos/Sources/RunnerForge/Models/ForgeConfig.swift
import Foundation

/// The contract between the GUI and every script: forge.json.
///
/// NO SECRET MAY EVER APPEAR IN THIS TYPE. It carries only non-secret
/// identifiers — app ids, installation ids, certificate common names, Azure
/// endpoint names. Every actual credential lives in the Keychain. A property
/// here that holds a password is a bug, not a feature.
public struct ForgeConfig: Codable, Sendable, Equatable {
    public var schemaVersion: Int
    public var github: GitHubConfig
    public var runners: [RunnerConfig]
    public var signing: SigningConfig
    public var paths: PathsConfig
    public var retention: RetentionConfig
    public var limits: LimitsConfig
    public var verification: VerificationConfig

    /// Runner Forge sends no telemetry. The property exists so the absence is
    /// explicit and auditable rather than merely unmentioned.
    public var telemetry: Bool

    public init(
        schemaVersion: Int = 1,
        github: GitHubConfig = GitHubConfig(),
        runners: [RunnerConfig] = [],
        signing: SigningConfig = SigningConfig(),
        paths: PathsConfig = PathsConfig(),
        retention: RetentionConfig = RetentionConfig(),
        limits: LimitsConfig = LimitsConfig(),
        verification: VerificationConfig = VerificationConfig(),
        telemetry: Bool = false
    ) {
        self.schemaVersion = schemaVersion
        self.github = github
        self.runners = runners
        self.signing = signing
        self.paths = paths
        self.retention = retention
        self.limits = limits
        self.verification = verification
        self.telemetry = telemetry
    }

    /// A config with every default filled in, for a first run.
    public static func makeDefault(workDir: String) -> ForgeConfig {
        ForgeConfig(
            runners: RunnerClass.all.map(RunnerConfig.init(forClass:)),
            paths: PathsConfig(workDir: workDir, aaxSdkSource: "")
        )
    }
}

public struct GitHubConfig: Codable, Sendable, Equatable {
    public var owner: String
    public var repos: [String]

    /// GitHub App id. NOT a secret; the private key lives in the Keychain.
    public var appId: String

    /// Installation id of the App on the owner account. NOT a secret.
    public var installationId: String

    public init(owner: String = "", repos: [String] = [], appId: String = "", installationId: String = "") {
        self.owner = owner
        self.repos = repos
        self.appId = appId
        self.installationId = installationId
    }
}

public struct RunnerConfig: Codable, Sendable, Equatable, Identifiable {
    public var classId: String
    public var enabled: Bool
    public var replicas: Int
    public var labels: [String]
    public var jobTimeoutMinutes: Int

    public var id: String { classId }

    public init(classId: String, enabled: Bool, replicas: Int, labels: [String], jobTimeoutMinutes: Int) {
        self.classId = classId
        self.enabled = enabled
        self.replicas = replicas
        self.labels = labels
        self.jobTimeoutMinutes = jobTimeoutMinutes
    }

    public init(forClass runnerClass: RunnerClass) {
        self.init(
            classId: runnerClass.classId,
            enabled: false,
            replicas: runnerClass.defaultReplicas,
            labels: runnerClass.labels,
            jobTimeoutMinutes: 60
        )
    }
}

public struct SigningConfig: Codable, Sendable, Equatable {
    /// Default is windows-ilok: the user's Windows PC runs 24/7.
    public var mode: String
    public var paceAccount: String
    public var paceWcGuid: String
    public var paceSignId: String

    /// Append --allowsigningservice to wraptool in cloud mode.
    public var allowSigningService: Bool
    public var windows: WindowsSigningConfig
    public var macos: MacosSigningConfig

    public var modeValue: SigningMode {
        get { SigningMode(rawValue: mode) ?? .windowsIlok }
        set { mode = newValue.rawValue }
    }

    public init(
        mode: String = SigningMode.windowsIlok.rawValue,
        paceAccount: String = "",
        paceWcGuid: String = "",
        paceSignId: String = "",
        allowSigningService: Bool = true,
        windows: WindowsSigningConfig = WindowsSigningConfig(),
        macos: MacosSigningConfig = MacosSigningConfig()
    ) {
        self.mode = mode
        self.paceAccount = paceAccount
        self.paceWcGuid = paceWcGuid
        self.paceSignId = paceSignId
        self.allowSigningService = allowSigningService
        self.windows = windows
        self.macos = macos
    }
}

public struct WindowsSigningConfig: Codable, Sendable, Equatable {
    public var provider: String
    public var azureEndpoint: String
    public var azureAccount: String
    public var azureProfile: String

    public init(
        provider: String = WindowsSigningProvider.azureTrustedSigning.rawValue,
        azureEndpoint: String = "",
        azureAccount: String = "",
        azureProfile: String = ""
    ) {
        self.provider = provider
        self.azureEndpoint = azureEndpoint
        self.azureAccount = azureAccount
        self.azureProfile = azureProfile
    }
}

public struct MacosSigningConfig: Codable, Sendable, Equatable {
    public var teamId: String

    /// A certificate common name, not a key. Safe to store here.
    public var devIdAppIdentity: String
    public var devIdInstallerIdentity: String

    /// Stapling follows notarization. Turning this off means the artifact fails
    /// Gatekeeper on any machine but the one that built it.
    public var notarize: Bool

    public init(
        teamId: String = "",
        devIdAppIdentity: String = "",
        devIdInstallerIdentity: String = "",
        notarize: Bool = true
    ) {
        self.teamId = teamId
        self.devIdAppIdentity = devIdAppIdentity
        self.devIdInstallerIdentity = devIdInstallerIdentity
        self.notarize = notarize
    }
}

public struct PathsConfig: Codable, Sendable, Equatable {
    public var workDir: String

    /// Local zip path or private git URL for the Avid AAX SDK. NEVER baked into
    /// an image; mounted from a local volume at build time. Empty means AAX
    /// targets are skipped and the workflow emits a marker naming the reason.
    public var aaxSdkSource: String

    public init(workDir: String = "", aaxSdkSource: String = "") {
        self.workDir = workDir
        self.aaxSdkSource = aaxSdkSource
    }
}

public struct RetentionConfig: Codable, Sendable, Equatable {
    /// Always true. Rebuilding base images costs tens of minutes.
    public var keepBaseImages: Bool

    /// Always true. Only per-job VM clones are ever deleted.
    public var keepTartImages: Bool

    public var keepCaches: Bool
    public var purgeOnExit: Bool
    public var logRetentionDays: Int

    public init(
        keepBaseImages: Bool = true,
        keepTartImages: Bool = true,
        keepCaches: Bool = true,
        purgeOnExit: Bool = true,
        logRetentionDays: Int = 7
    ) {
        self.keepBaseImages = keepBaseImages
        self.keepTartImages = keepTartImages
        self.keepCaches = keepCaches
        self.purgeOnExit = purgeOnExit
        self.logRetentionDays = logRetentionDays
    }
}

public struct LimitsConfig: Codable, Sendable, Equatable {
    public var maxDiskGb: Int
    public init(maxDiskGb: Int = 120) { self.maxDiskGb = maxDiskGb }
}

public struct VerificationConfig: Codable, Sendable, Equatable {
    public var canaryRepo: String

    /// Consecutive green runs required before the setup counts as stable.
    public var requiredGreenRuns: Int

    /// Written by SelfTestService. Nil until the first self-test.
    public var lastProof: SelfTestResult?

    public init(canaryRepo: String = "", requiredGreenRuns: Int = 3, lastProof: SelfTestResult? = nil) {
        self.canaryRepo = canaryRepo
        self.requiredGreenRuns = requiredGreenRuns
        self.lastProof = lastProof
    }
}

File: app-windows/RunnerForge/Models/ForgeConfig.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;

namespace RunnerForge.Models;

/// <summary>
/// The contract between the GUI and every script: forge.json.
/// </summary>
/// <remarks>
/// NO SECRET MAY EVER APPEAR IN THIS TYPE. It carries only non-secret
/// identifiers — app ids, installation ids, certificate common names, Azure
/// endpoint names. Every actual credential lives in the OS keystore. A property
/// here that holds a password is a bug, not a feature.
/// </remarks>
public sealed class ForgeConfig
{
    [JsonPropertyName("schemaVersion")]
    public int SchemaVersion { get; set; } = 1;

    [JsonPropertyName("github")]
    public GitHubConfig GitHub { get; set; } = new();

    [JsonPropertyName("runners")]
    public List<RunnerConfig> Runners { get; set; } = [];

    [JsonPropertyName("signing")]
    public SigningConfig Signing { get; set; } = new();

    [JsonPropertyName("paths")]
    public PathsConfig Paths { get; set; } = new();

    [JsonPropertyName("retention")]
    public RetentionConfig Retention { get; set; } = new();

    [JsonPropertyName("limits")]
    public LimitsConfig Limits { get; set; } = new();

    [JsonPropertyName("verification")]
    public VerificationConfig Verification { get; set; } = new();

    /// <summary>
    /// Runner Forge sends no telemetry. The property exists so the absence is
    /// explicit and auditable rather than merely unmentioned.
    /// </summary>
    [JsonPropertyName("telemetry")]
    public bool Telemetry { get; set; }

    public static JsonSerializerOptions SerializerOptions { get; } = new()
    {
        WriteIndented = true,
        DefaultIgnoreCondition = JsonIgnoreCondition.Never,
    };

    /// <summary>A config with every default filled in, for a first run.</summary>
    public static ForgeConfig CreateDefault(string workDir) => new()
    {
        SchemaVersion = 1,
        Runners = RunnerClass.All.Select(RunnerConfig.FromClass).ToList(),
        Paths = new PathsConfig { WorkDir = workDir, AaxSdkSource = "" },
        Telemetry = false,
    };
}

public sealed class GitHubConfig
{
    [JsonPropertyName("owner")] public string Owner { get; set; } = "";
    [JsonPropertyName("repos")] public List<string> Repos { get; set; } = [];

    /// <summary>GitHub App id. NOT a secret; the private key lives in the keystore.</summary>
    [JsonPropertyName("appId")] public string AppId { get; set; } = "";

    /// <summary>Installation id of the App on the owner account. NOT a secret.</summary>
    [JsonPropertyName("installationId")] public string InstallationId { get; set; } = "";
}

public sealed class RunnerConfig
{
    [JsonPropertyName("classId")] public string ClassId { get; set; } = "";
    [JsonPropertyName("enabled")] public bool Enabled { get; set; }
    [JsonPropertyName("replicas")] public int Replicas { get; set; } = 1;
    [JsonPropertyName("labels")] public List<string> Labels { get; set; } = [];
    [JsonPropertyName("jobTimeoutMinutes")] public int JobTimeoutMinutes { get; set; } = 60;

    public static RunnerConfig FromClass(RunnerClass runnerClass) => new()
    {
        ClassId = runnerClass.ClassId,
        Enabled = false,
        Replicas = runnerClass.DefaultReplicas,
        Labels = [.. runnerClass.Labels],
        JobTimeoutMinutes = 60,
    };
}

public sealed class SigningConfig
{
    /// <summary>Default is windows-ilok: the user's Windows PC runs 24/7.</summary>
    [JsonPropertyName("mode")] public string Mode { get; set; } = "windows-ilok";

    [JsonPropertyName("paceAccount")] public string PaceAccount { get; set; } = "";
    [JsonPropertyName("paceWcGuid")] public string PaceWcGuid { get; set; } = "";
    [JsonPropertyName("paceSignId")] public string PaceSignId { get; set; } = "";

    /// <summary>Append --allowsigningservice to wraptool in cloud mode.</summary>
    [JsonPropertyName("allowSigningService")] public bool AllowSigningService { get; set; } = true;

    [JsonPropertyName("windows")] public WindowsSigningConfig Windows { get; set; } = new();
    [JsonPropertyName("macos")] public MacosSigningConfig Macos { get; set; } = new();

    [JsonIgnore]
    public SigningMode ModeEnum
    {
        get => SigningModeExtensions.FromConfigValue(Mode);
        set => Mode = value.ToConfigValue();
    }
}

public sealed class WindowsSigningConfig
{
    [JsonPropertyName("provider")] public string Provider { get; set; } = "azure-trusted-signing";
    [JsonPropertyName("azureEndpoint")] public string AzureEndpoint { get; set; } = "";
    [JsonPropertyName("azureAccount")] public string AzureAccount { get; set; } = "";
    [JsonPropertyName("azureProfile")] public string AzureProfile { get; set; } = "";
}

public sealed class MacosSigningConfig
{
    [JsonPropertyName("teamId")] public string TeamId { get; set; } = "";

    /// <summary>A certificate common name, not a key. Safe to store here.</summary>
    [JsonPropertyName("devIdAppIdentity")] public string DevIdAppIdentity { get; set; } = "";

    [JsonPropertyName("devIdInstallerIdentity")] public string DevIdInstallerIdentity { get; set; } = "";
    [JsonPropertyName("notarize")] public bool Notarize { get; set; } = true;
}

public sealed class PathsConfig
{
    [JsonPropertyName("workDir")] public string WorkDir { get; set; } = "";

    /// <summary>
    /// Local zip path or private git URL for the Avid AAX SDK. NEVER baked into
    /// an image layer; mounted from a local named volume at build time. Empty
    /// means AAX targets are skipped and the workflow emits a marker artifact
    /// naming the reason.
    /// </summary>
    [JsonPropertyName("aaxSdkSource")] public string AaxSdkSource { get; set; } = "";
}

public sealed class RetentionConfig
{
    /// <summary>Always true. Rebuilding base images costs tens of minutes.</summary>
    [JsonPropertyName("keepBaseImages")] public bool KeepBaseImages { get; set; } = true;

    /// <summary>Always true. Only per-job VM clones are ever deleted.</summary>
    [JsonPropertyName("keepTartImages")] public bool KeepTartImages { get; set; } = true;

    [JsonPropertyName("keepCaches")] public bool KeepCaches { get; set; } = true;
    [JsonPropertyName("purgeOnExit")] public bool PurgeOnExit { get; set; } = true;
    [JsonPropertyName("logRetentionDays")] public int LogRetentionDays { get; set; } = 7;
}

public sealed class LimitsConfig
{
    [JsonPropertyName("maxDiskGb")] public int MaxDiskGb { get; set; } = 120;
}

public sealed class VerificationConfig
{
    [JsonPropertyName("canaryRepo")] public string CanaryRepo { get; set; } = "";

    /// <summary>Consecutive green runs required before the setup counts as stable.</summary>
    [JsonPropertyName("requiredGreenRuns")] public int RequiredGreenRuns { get; set; } = 3;

    /// <summary>Written by SelfTestService. Null until the first self-test.</summary>
    [JsonPropertyName("lastProof")] public SelfTestResult? LastProof { get; set; }
}
